scripts/powershell/copy-secrets.ps1

```diff
@@ -0,0 +1,133 @@
+  .\Copy-AzKeyVaultSecret.ps1 -SourceSubscriptionId <String> -TargetSubscriptionId <String> -SourceVaultName <String> -TargetVaultName <String>
+#>
+
+param (
+  [Parameter(Mandatory = $true)]
+  [string]$SourceSubscriptionId,
+
+  [Parameter(Mandatory = $false)]
+  [string]$TargetSubscriptionId,
+
+  [Parameter(Mandatory = $true)]
+  [string]$SourceVaultName,
+
+  [Parameter(Mandatory = $true)]
+  [string]$TargetVaultName,
+
+  # Force overwrite existing secrets
+  [Parameter(Mandatory = $false)]
+  [switch]$Force
+)
+
+$IpAddress = (Invoke-RestMethod -Uri "https://api.ipify.org")
+$IpAddressRange = "$IpAddress/32"
+Write-Information "Current IP address: $IpAddress"
+
+$Context = Set-AzContext -Subscription $SourceSubscriptionId
+Write-Information "Current subscription: $($Context.Subscription.Name)"
+
+$SourceVault = Get-AzKeyVault -VaultName $SourceVaultName
+$AddNetworkRule = $SourceVault.NetworkAcls.IpAddressRanges -notcontains $IpAddressRange
+
+try {
+  # Add IP adderss to source Key vault
+  if ($AddNetworkRule) {
+    $null = Add-AzKeyVaultNetworkRule -VaultName $SourceVaultName -IpAddressRange $IpAddressRange
+  }
+
+  # Get list of secrets names
+  $SourceVaultSecretNames = (Get-AzKeyVaultSecret -VaultName $SourceVaultName).Name
+
+  $SourceVaultSecrets = @()
+  $SourceVaultSecretNames | ForEach-Object {
+    $SourceVaultSecrets += (Get-AzKeyVaultSecret -VaultName $SourceVaultName -Name $_)
+  }
+}
+catch {
+  Write-Error "An error occurred: $_"
+}
+finally {
+  # Remove IP address from source Key vault
+  if ($AddNetworkRule) {
+    $null = Remove-AzKeyVaultNetworkRule -VaultName $SourceVaultName -IpAddressRange $IpAddressRange
+  }
+}
+
+# Switch context to target subscription if $TargetSubscriptionId has value
+if (![string]::IsNullOrEmpty($TargetSubscriptionId)) {
+  $Context = Set-AzContext -Subscription $TargetSubscriptionId
+  Write-Information "Target subscription: $($Context.Subscription.Name)"
+}
+
+$TargetVault = Get-AzKeyVault -VaultName $TargetVaultName
+$AddNetworkRule = $TargetVault.NetworkAcls.IpAddressRanges -notcontains $IpAddressRange
+
+try {
+  # Add IP address to target Key vault
+  if ($AddNetworkRule) {
+    $null = Add-AzKeyVaultNetworkRule -VaultName $TargetVaultName -IpAddressRange $IpAddressRange
+  }
+
+  # Copy secrets
+  $SourceVaultSecrets | ForEach-Object {
+    $SourceVaultSecretName    = $($_.Name)    # Fetch secret name
+    $SourceVaultSecretExpDate = $($_.Expires) # Fetch expiration date
+    $TargetVaultSecret        = Get-AzKeyVaultSecret -VaultName $TargetVaultName -Name $SourceVaultSecretName
+    $SecretValue              = $($_.SecretValue)
+
+    # Skip if secret exists in target vault
+    if ($TargetVaultSecret -eq $null -or $Force) {
+      # Secret does not exist
+      $Copy = Set-AzKeyVaultSecret -VaultName $TargetVaultName -Name $_.Name -Expires $SourceVaultSecretExpDate -SecretValue $SecretValue
+      Write-Output "Successfully replicated secret '$($Copy.Id)'"
```

Comment:
- Rename `$Source*` variables to `$Target*`, since `$_` already represents the source secret and these variables should represent properties of the target secret.
- Reorder variable declarations.
- Move somve variable declarations into if-statement, since they're only used inside that statement. No need to declare them unless needed.
- Don't wrap variable values in `$()` if not needed.

```suggestion
    $TargetVaultSecretName    = $_.Name
    $TargetVaultSecret        = Get-AzKeyVaultSecret -VaultName $TargetVaultName -Name $TargetVaultSecretName

    # Skip if secret exists in target vault
    if ($TargetVaultSecret -eq $null -or $Force) {
      # Secret does not exist
      $TargetVaultSecretExpDate = $_.Expires
      $TargetSecretValue              = $_.SecretValue
      $Copy = Set-AzKeyVaultSecret -VaultName $TargetVaultName -Name $TargetVaultSecretName -Expires $TargetVaultSecretExpDate -SecretValue $SecretValue
      Write-Output "Successfully replicated secret '$($Copy.Id)'"
```

Reply:
@chbul This turned out pretty messy. Let's treat this comment as my notes and have a look together 😄

Reply:
Nice discussion point for next "parprogrammering"! 👍 Here I used `$()` intentionally with `$PSItem` (or `$_`) in pipeline loops when assigning `$_.Name` to a new variable because they have expandable/interpolating strings with an [expression inside to be evaluated](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_quoting_rules?view=powershell-7.4#double-quoted-strings):

`$MyName = $($_.Name)`

Reply:
> * Don't wrap variable values in `$()` if not needed.

Great discussion point for next "parprogrammering"! 👍 Here I used `$()` intentionally with `$PSItem` (or `$_`) in the pipeline loop when assigning `$_.Name` to a new variable because they have expandable/interpolating strings with an [expression inside to be evaluated](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_quoting_rules?view=powershell-7.4#double-quoted-strings):

`$MyName = $($_.Name)`